Merge Speed up and Refactoring

kim-mskw opened the pull request:
This pull request introduces several updates and improvements across multiple data pipeline and training scripts, primarily to enhance configuration flexibility, streamline data handling, and improve model training workflows. The most significant changes include consistent use of start/end date configuration, new data preprocessing and postprocessing steps, and enhancements to the model training script to support resuming from checkpoints and improved logging.

**Configuration and Data Handling Updates:**

- Standardized the use of `DATA_START` and `DATA_END` variables (from `src.shared.config`) in place of the previous `START` and `END` throughout data acquisition and transformation scripts, ensuring a consistent and configurable approach to specifying data ranges. 

- Improved environment variable loading by using `find_dotenv()` and setting `override=True` to ensure correct environment configuration.

**Data Processing and Preprocessing Enhancements:**

- Added new postprocessing steps in `01b_data_acquisition.py`:
    - Rebuilt and merged ID-full data using `run_build_idfull_and_merge`.
    - Introduced a precompute step for VWAP matrices (required for downstream algorithms) with `precompute_range`, saving results to a configurable path. 

- Updated data transformation logic to use dynamically generated file names based on configuration dates and removed hardcoded date ranges. 

**Simulation and Model Training Improvements:**

- Modified simulation scripts to reference new or updated simulation functions, such as switching to `simulate_period` from `rolling_intrinsic_gurobi_qh` and updating the multi-market simulation to use `simulate_days_stacked_quarterhourly_products`. 

- Enhanced the training script for coordinated multi-market models:
    - Added support for resuming training from a specified checkpoint, including logic to handle versioned output directories.
    - Improved device selection and logging.
    - Updated model instantiation to include new parameters (e.g., `reward_log_path`) and adjusted hyperparameters.
    - Incorporated more robust handling of input data splits and preprocessing. 

**Other Notable Changes:**

- Cleaned up and clarified code, such as removing unused variables, commenting out legacy code, and improving comments for maintainability. 

These updates collectively improve the maintainability, configurability, and robustness of the data pipeline and model training workflow.

TimFsl committed:
moved position of load_dotenv to start of script

TimFsl committed:
changes in load_entsoe function

TimFsl committed:
fixes in etl_transactions_new_format

TimFsl committed:
deleted file

TimFsl committed:
changed day ahead milp solver to gurobi

TimFsl committed:
changed dates until 01.01.2024

TimFsl committed:
added ID Full data to csv file by calcualting it based on the epex transactions in the database

TimFsl committed:
normalized datetime index

TimFsl committed:
ensured that time step is within bounds

TimFsl committed:
adjusteddatabase connetion in RI script

TimFsl committed:
copy of RI script and changes to optimization function for testing

TimFsl committed:
calling new adjusted RI model

TimFsl committed:
test ri

TimFsl committed:
testing differnet RI versions

TimFsl committed:
new versions of RI with precomputed VWAPSs as parquet files

TimFsl committed:
adjsuted RI intelligent stacking be be able to handle precomputed vwaps

TimFsl committed:
adjusted RI test stacking

TimFsl committed:
fixed multiple bugs

TimFsl committed:
adjusted resiudal load calucaltion in environment

TimFsl committed:
new training, validation and test splitting

TimFsl committed:
adjusted scaling of day ahead reward signal in step function, adjsuted "check if complete cycle" function so RI reward is added even when no complete DA cycle

TimFsl committed:
adjusted scaler

TimFsl committed:
mostly changes in etl pipeline

TimFsl committed:
new gurobipy RI model

TimFsl committed:
bug fixes in custom_ppo and new RI script

TimFsl committed:
changes in drl pipeline

TimFsl committed:
tested differnet batch sizes and clipping rates

TimFsl committed:
changed reuired steps per epidsode to 24

TimFsl committed:
cleaned repository, added comments and dynamic references

TimFsl committed:
changed names of RI scripts and moved old scripts into archive folder

TimFsl committed:
fixed bug in loading forecast data

TimFsl committed:
adjusted the entsoe api call to prevent hitting request limit

TimFsl committed:
added function to calculate vwap matrices

kim-mskw committed:
- some cleaning of old code added TODO with comments about the changes needed

kim-mskw committed:
- alle meine Kommentare als TODO geadded weil der Review ist ultra weird gerade

TimFsl commented on 01b_data_acquisition.py:
```diff
@@ -115,7 +115,7 @@
 data_hourly.to_csv(csv_path)
 
-
+# TODO: Das brauchst du hier nciht korrekt?
```
Ja das kann raus

TimFsl commented on 01b_data_acquisition.py:
```diff
@@ -125,6 +125,7 @@
 # You can adjust bucket size and min trades as needed
 # Matrices are required for Rolling Intrinsic Algortihm
 
+#TODO: Hier nicht ordentlich die config benutzt und das macht keinen Sinn das als Funktion zu definieren. 
```
ja das stimmt. Ich hab das angepasst in config und dem runner und unter main in precompute_vwaps die Importe aus config hinzugefügt falls man nur das script laufen lassen will

TimFsl commented on 04b_train_coordinated_multi_market.py:
```diff
@@ -147,6 +147,7 @@
             n_steps=512,
             clip_range=0.4,
             batch_size=128,
+            # TODO: wieso hast du das Gewicht von dem value loss kleiner gemacht?
```
Ich habe mit den ganzen Parametern rumgespielt und getestet, der Wert hat keine spezielle Bedeutung

TimFsl commented on 02b_single_market_rolling_intrinsic_qh.py:
```diff
@@ -19,7 +19,7 @@
 
 if __name__ == "__main__":
-
+# TODO: Du hattest vorher die Bucket size in dem vWAp pre calculation neu gesetzt. man sollte hier keinen Möglichkeit haben die zu überschrieben sondern sich die aus dem precalculated file holen. Überlege mal ob du das alt meta da mit ablegen kannst. Ansonsten führt das nur zu komischen Fehlern. 
```
Ja ergibt Sinn. Ich habe jetzt bei allen drei RI Skripten eine neue mini Funktion eingebaut welche aus den VWAPs Matrizen über die Indizes die Größe ableitet. Die wird dann innerhalb der ersten while Schleife aufgerufen und so die bucket size bestimmt.

TimFsl commented on src/coordinated_multi_market/rolling_intrinsic/testing_rolling_intrinsic_qh_intelligent_stacking.py:
```diff
@@ -190,6 +190,7 @@ def build_battery_model(
     c_rate: float,
     roundtrip_eff: float,
 ) -> Tuple[gp.Model, Dict[str, Any], Dict[Any, gp.Constr], gp.Constr]:
+    # TODO: gleiches hier hast du verifiziert, dass das Model vorher und nachher gleich funktioniert hat?
```
Ja, die Modelle liefern bis auf minimale numerische Abweichungen dieselben Ergebnisse.

TimFsl commented on src/coordinated_multi_market/basic_battery_dam_env.py:
```diff
@@ -182,6 +182,7 @@ def step(
             # check if we have capacity left in battery
             if self._current_soc > 0:
                 # penalty because missed profit
+                # TODO: Ahhhh, ja wie sieht deine penalty im Vergleich zum Reward denn jetzt aus? Macht ja Sinn das der sich anders verhält im Vergleich zu meinem Training. Deine peanlty ist jetzt ja viel kleiner. 
```
Ja hier gilt dasselbe wie bei den Parametern, ich habe mit dem Reward Sharing rumgespielt und Sachen ausprobiert. Besser noch nicht übernommen sondern einfach einmal final wenn das Modell sauber läuft

TimFsl commented on src/coordinated_multi_market/learning_utils.py:
```diff
@@ -13,6 +13,7 @@
 train_end = DATA_END.date().isoformat()
 path = f"df_spot_train_{train_start}_{train_end}_with_features_utc.csv"
 
+# TODO: Finde ich nicht gut. Würde dir empfehlen auf das alte Sampling zu ändern.
```
Lass uns dazu am Montag nochmal reden

TimFsl commented on src/data_acquisition/epex_sftp/build_idfull.py:
```diff
@@ -110,7 +110,7 @@ def merge_idfull_into_csv(csv_in: str, csv_out: str = None) -> pd.DataFrame:
     if "id_full_h" in df.columns:
         df.drop(columns=["id_full_h"], inplace=True)
-        
+    #TODO: Benutzen das später für den Spread. Evt kann man da noch andere Spreads nehmen. Zum Beispiel den ID3 für das Produkt, weil das wird häufig als proxy für den Kauf/Verkauf auf dem ID genommen.  
```
Verstehe den Kommentar nicht ganz. Meinst du ich soll auch noch die anderen Indizes berechnen wie ID3 ?

TimFsl commented on src/coordinated_multi_market/custom_ppo.py:
```diff
@@ -177,6 +168,7 @@ def collect_rollouts(
             # --- DA-Teil: immer vorhanden ---
             da_rewards = rollout_buffer.rewards[row_start : row_start + num_rows].flatten()
 
+            # TODO: Wenn du es immer berechnest warum ist es dann optional?
```
Ja der Kommentar kann weg

TimFsl commented on src/coordinated_multi_market/custom_ppo.py:
```diff
@@ -422,12 +411,10 @@ def _check_if_complete_cycle(period_volumes, capacity: float = 1.0,
         traded_volume = abs(period_volumes).sum()
         cycle_fraction = traded_volume / (2 * capacity)
+        
+        # TODO: Wieso hier größer gleich? IN welchem Case erlaubst du mehr als 1 Cycle?
```
Das ganze ist nur relevant wenn man den Wert von min_cycle_fraction auf kleiner als 1 setzt. Daher das größer gleich Operator. Ich hatte die Funktion umgeändert damit es möglich ist dass der Agent eine offene Position eingehen kann und trotzdem RI-Reward bekommt. Wenn das auf 1 ist muss er ja einen vollständigen Cycle im DayAhead machen um überhaupt RI-Reward zu bekommen. Wenn man den min_cycle_fraction Wert auf etwas zwischen 0 bis 1 setzt hat er dann mehr Flexibilität

TimFsl commented on src/single_market/rolling_intrinsic_gurobi_qh.py:
```diff
@@ -562,6 +562,7 @@ def simulate_period(
         logger.info(f"Trading end:   {trading_end}")
 
         # Battery time index (delivery quarter-hours)
+        # TODO: warum so kompliziert?
```
habe ich angepasst.

TimFsl commented on src/data_acquisition/epex_sftp/precompute_vwaps.py:
```diff
@@ -129,17 +126,19 @@ def precompute_vwaps_for_day(
     )
 
     # All quarter-hourly products for this delivery day
+    # TODO: Warum ist das so umständlich definiert? start und end of day sind doch eh fixe Werte? 
```
Habe ich angepasst

TimFsl committed:
Fix VWAP precompute day normalization & index generation (DST-safe)

Normalize current_day to Europe/Berlin midnight (tz_convert + normalize)
Adjust bucket_index to align with bucket boundaries (start at trading_start, exclude last incomplete bucket)
Simplify product_index generation to full 15-min grid of the delivery day (inclusive="left")
Rename clarity: volume_weighted_price → trade_value
Add bucket_start column for easier debugging/analysis (unused in aggregation)

TimFsl committed:
normalize delivery-day indexing and timezone handling

Convert price discounting time deltas in UTC to avoid DST/timezone drift
Standardize delivery-day QH index to [00:00, 24:00) using normalize()+inclusive="left"
Infer bucket size from VWAP matrix index spacing (remove explicit bucket_size param)
Simplify battery time index generation and normalize current_day

TimFsl committed:
derive bucket size from VWAP matrix

TimFsl committed:
derive bucket size from VWAP matrix

TimFsl committed:
chnged comments to english

TimFsl committed:
Derive and add daily EXAA forecast statistics as additional features

Compute daily stats (mean/std/min/max/spread) from exaa_15min series
Add intra-day ramp features (abs diff sum/max)
Include new EXAA daily features in df_spot_train export

TimFsl committed:
extend scaled feature set with EXAA daily stats and update day-shape validation

TimFsl committed:
curriculum cycles + richer obs; refactor step/reward + debug logging

Introduce max_cycles (curriculum learning) and episode/global step tracking
Add EXAA daily features; update observation size 43→50
Map forecast price scaling to [-1,1] to match observation Box
Enforce feasible actions w/ overflow penalty; normalize reward and adjust terminal SOC penalty
Add debug CSV logging per step (obs vector + timestamps)

TimFsl committed:
Refactor PPO rollout: curriculum learning, delayed IDC rewards & episodic logging

Introduced curriculum learning for max_cycles based on training steps
Delayed intraday (IDC) reward activation via configurable START_IDC_STEPS
Changed IDC reward shaping: single terminal reward per episode instead of per-step spreading
Simplified complete cycle check (allow partial cycles early in training)
Refactored episodic reward & profit logging with aggregated TensorBoard metrics
Improved rollout buffer handling and removed duplicated reward overwrites
Aligned buffer sizes with n_rollout_steps for safety and clarity

kim-mskw committed:
- change db documentation
- extend usage von env variables including password and port
- test milp db and data fill stuff

kim-mskw committed:
- consistency of function calls
- update readme to mention gurobi license
- update requirements txt
- consitently use thesisdb hook to avoid double hard cofing of sql url

kim-mskw committed:
- config changes for dynmic path recognition in all file names
- reset the sampling to random sampels

kim-mskw committed:
- reset some learning changes, NOT TESTED
